client/dfclient/src/loading_state.cpp

```diff
@@ -0,0 +1,164 @@
+	_resources.textures["fillup_outline"] = std::make_unique<ncine::Texture>((rootPath + "loading/outline.png").c_str());
+	_resources.textures["blackpixel.png"] = std::make_unique<ncine::Texture>((rootPath + "loading/blackpixel.png").c_str());
+
+	// init loading text
+	_loadingText = std::make_unique<ncine::TextNode>(&rootNode, _resources.fonts["comic"].get());
+	_loadingText->setColor(ncine::Color::White);
+	_loadingText->setPosition(res.x*0.5f, res.y*0.5f);
+	_loadingText->setString("Loading... 0%");
+
+	// init loading graphics
+	_fillupSprite = std::make_unique<ncine::Sprite>(&rootNode, _resources.textures["fillup"].get());
+	_fillupSprite->setPosition(res.x*0.5f, res.y*0.6f);
+	_fillupSprite->setLayer(1);
+
+	_curtain = std::make_unique<ncine::MeshSprite>(&rootNode, _resources.textures["blackpixel.png"].get());
+	std::vector<ncine::Vector2f> vertices(4);	
+	vertices[0].x = _fillupSprite->position().x - _fillupSprite->width()/2;
+	vertices[0].y = _fillupSprite->position().y - _fillupSprite->height()/2;
+	vertices[1].x = _fillupSprite->position().x + _fillupSprite->width()/2;
+	vertices[1].y = _fillupSprite->position().y - _fillupSprite->height()/2;
+	vertices[2].x = _fillupSprite->position().x - _fillupSprite->width()/2;
+	vertices[2].y = _fillupSprite->position().y + _fillupSprite->height()/2;
+	vertices[3].x = _fillupSprite->position().x + _fillupSprite->width()/2;
+	vertices[3].y = _fillupSprite->position().y + _fillupSprite->height()/2;
+	_curtain->createVerticesFromTexels(vertices.size(), vertices.data());
+	_curtain->setPosition(0, 0);
+	_curtain->setLayer(2);
+
+	_outline = std::make_unique<ncine::Sprite>(&rootNode, _resources.textures["fillup_outline"].get());
+	_outline->setPosition(res.x*0.5f, res.y*0.6f);
+	_outline->setLayer(3);
+}
+
+StateType LoadingState::Update(Messages m) {
+	// load a single file
+	std::string path = "";
+	if (!_textureFilenames.empty()){
+		path = _textureFilenames.back();
+		std::string filename = ncine::FileSystem::baseName(path.c_str()).data();
+		_resources.textures[filename] = std::make_unique<ncine::Texture>(path.c_str());
+		_textureFilenames.pop_back();
+
+	} else if (!_soundFilenames.empty()) {
+		path = _soundFilenames.back();
+		std::string filename = ncine::FileSystem::baseName(path.c_str()).data();
+		_resources._sounds[filename] = std::make_unique<ncine::AudioBuffer>(path.c_str());
+		_soundFilenames.pop_back();
+	}
+	_current += fileSize(path);
+
+	// update the loading text
+	auto _percent_loaded = _current/(double)_total * 100;
+	auto string = "Loading... " + std::to_string((int)_percent_loaded) + "%";
+	_loadingText->setString(string.c_str());
+	unsigned char alpha = 55 + 2*_percent_loaded;
+	_loadingText->setAlpha(alpha);
+
+	// update the loading graphics
+	_curtain->setPosition(0, 1.2*_percent_loaded);
+
+	// manage the ending
+	if (_percent_loaded == 100) {
+		if (_fadeoutTween == nullptr) {
+			// stretch curtain to cover the whole screen
```

Comment:
to be honest I'm not sure whether I'm seeing the curtain effect at all. but overall the game looks much better and doesn't crash so that's ok :D

Reply:
The curtain then is used to do the fade-out effect, it's a black rectangle that is becoming less and less transparent. It doesn't go anywhere after that, just disappears :D